feature: Add AWS MSK IAM Authentication Support for Kafka Engine

kalavt opened the pull request:
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes into CHANGELOG.md):
Add support for AWS MSK IAM authentication for Kafka.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### Description

This PR implements AWS MSK IAM authentication support for the Kafka Table Engine.
It allows ClickHouse to authenticate with Amazon Managed Streaming for Apache Kafka (MSK) using IAM roles.

**Motivation:**
Many users running ClickHouse on AWS want to connect to MSK using the standard IAM authentication method for better security and management, rather than managing SASL/SCRAM credentials or mTLS certificates.

**Changes:**
- Added `AWSMSKIAMAuth` class (`src/Storages/Kafka/AWSMSKIAMAuth.cpp/h`) to handle IAM authentication logic.
- Updated `StorageKafka` and `KafkaConsumer` to use the new authentication mechanism.
- Added configuration settings for AWS MSK IAM auth.
- Updated `librdkafka-cmake` to support the necessary build options.
- Updated documentation with usage examples.

antaljanosbenjamin reviewed (commented):
In general looks good. There are a few general comments:
1. Could you please also update the regarding docs?
2. Have you thought about using "assumed roles" similarly to the S3 table function? I think it is implemented for S3, it shouldn't be hard to enable it here too.
3. I have to talk with some security people, but we might need a setting to disable this globally if there is some issue regarding security. Please don't act on this yet, I will try to get a precise answer on that and understand it better.

kalavt commented on src/Storages/Kafka/KafkaSettings.cpp:
```diff
@@ -45,7 +45,7 @@ namespace ErrorCodes
     DECLARE(String, kafka_replica_name, "", "The replica name in ClickHouse Keeper", 0) \
     DECLARE(String, kafka_security_protocol, "", "Protocol used to communicate with brokers.", 0) \
-    DECLARE(String, kafka_sasl_mechanism, "", "SASL mechanism to use for authentication. Supported: GSSAPI, PLAIN, SCRAM-SHA-256, SCRAM-SHA-512, OAUTHBEARER.", 0) \
+    DECLARE(String, kafka_sasl_mechanism, "", "SASL mechanism to use for authentication. Supported: GSSAPI, PLAIN, SCRAM-SHA-256, SCRAM-SHA-512, OAUTHBEARER, aws_msk_iam.", 0) \
```
Thanks for point out, fixed.

kalavt commented on src/Storages/Kafka/KafkaConfigLoader.cpp:
```diff
@@ -364,7 +365,33 @@ void updateConfigurationFromConfig(
+        String sasl_mechanism = kafka_settings[KafkaSetting::kafka_sasl_mechanism].value;
+        
+        // Check if this is AWS MSK IAM authentication
+        if (sasl_mechanism == "aws_msk_iam")
```
I've updated the code to use case-insensitive comparison with `boost::iequals()` to match librdkafka's behavior

ilejn commented on src/Storages/Kafka/AWSMSKIAMAuth.cpp:
```diff
@@ -0,0 +1,331 @@
+    // Detect cluster type by checking for serverless keyword in broker address
+    bool is_serverless = false;
+    if (!broker_list.empty() && broker_list.find(".kafka-serverless.") != String::npos)
+    {
```
broker_list.contains(..)

antaljanosbenjamin commented:
The changes in c-ares are intentional?

kalavt commented:
@antaljanosbenjamin 

Regarding the use_environment_credentials flag — I believe this parameter might not be necessary in this context.

For Kafka IAM authentication to succeed, the AWS IAM principal must already have explicit permissions such as:
	•	kafka-cluster:Connect
	•	and the relevant topic-level permissions

Without these permissions, authentication/authorization will fail regardless of how credentials are sourced.
Given this, enabling or disabling environment credentials at the ClickHouse server layer doesn’t change the security model — IAM permissions still enforce the boundary. Therefore, adding an additional use_environment_credentials switch may not bring meaningful security benefits, while adding extra complexity both for configuration and code paths.

Let me know if you see a specific scenario where restricting environment credentials at the ClickHouse level improves security beyond what IAM already enforces. I’m happy to adjust if there’s a concrete threat model or requirement I might be missing.

antaljanosbenjamin commented on src/Storages/Kafka/KafkaConfigLoader.cpp:
```diff
@@ -160,6 +170,15 @@ namespace
+    static const std::unordered_set<String> clickhouse_internal_settings = {
+        "use_environment_credentials",  // AWS MSK IAM: controls credential source (env vs instance profile)
+        // Add future internal settings here
+    };
```
I like the general approach (have a list of internal settings that we ignore), but let's ignore them in `loadFromConfig`, similarly how `CONFIG_KAFKA_PRODUCER_TAG` and `CONFIG_KAFKA_CONSUMER_TAG` are ignored.

antaljanosbenjamin reviewed (commented):
A few small things.

kalavt commented on src/Storages/Kafka/AWSMSKIAMAuth.cpp:
```diff
@@ -0,0 +1,415 @@
+        const char * /* oauthbearer_config */,
+        void * opaque)
+    {
+        auto log = getLogger("AWSMSKIAMAuth");
```
Agreed and implemented! The logger is now stored in the `OAuthBearerTokenRefreshContext` structure (see AWSMSKIAMAuth.h).

antaljanosbenjamin replied:
This is still here.

kalavt commented:
@antaljanosbenjamin can you pls guide me for next step merge this PR?

antaljanosbenjamin commented on src/Storages/Kafka/AWSMSKIAMAuth.cpp:
```diff
@@ -0,0 +1,277 @@
+                        aws_client_configuration, Aws::Auth::AWSCredentials{}, credentials_configuration);
+                }
+                provider = g_provider_cache.provider;
+            }
```
What happens if the region is different in the current config then it was when the cache was created?

I might miss something, but I don't understand why the storage cannot own the credentials provider. it is tight to a single storage, clear ownership and scope. Now we have a global object which is shared among multiple storages.

kalavt committed:
feat: Add AWS MSK IAM authentication support for Kafka

- Implements AWS MSK IAM authentication for Kafka storage.
- Updates librdkafka-cmake configuration.

antaljanosbenjamin reviewed (changes requested):
A few comments, please don't use force push.

antaljanosbenjamin commented on src/Storages/Kafka/AWSMSKIAMAuth.cpp:
```diff
@@ -0,0 +1,306 @@
+        const char * oauthbearer_config,
+        void * /* opaque */)
+    {
+        LoggerPtr log = getLogger("AWSMSKIAMAuth");
```
And please, remove this log.

kalavt commented on src/Storages/Kafka/KafkaConfigLoader.cpp:
```diff
@@ -364,7 +381,33 @@ void updateConfigurationFromConfig(
+                : "";
+
+            AWSMSKIAMAuth::setupAuthentication(
+                kafka_config, params.config, aws_region, broker_list, params.log, storage.getOAuthContext());
```
You're correct. The OAuth context initialization is expensive and should be cached. I'll refactor this part.

kalavt committed:
feat: AWS MSK IAM fix review comments
